### PR TITLE
use FETCH_HEAD timestamp to avoid excessive background fetching

### LINKED ISSUE #3329: Don't always fetch after switching repo's and/or fetch less often
Hi all, a small suggestion for an improvement; 

Currently, every time you switch to a different repository, the first thing the app does is fetch. It does so, even though it may have just done it 5 seconds ago for that same repository.

Would it be possible to keep track of when an (auto) fetch was last done for a repository, and then avoid doing it too often? For example no more than once every minute.

I switch back and forth between different repositories a lot, and this makes the app fetch over and over again, quite redundantly. I think this is bad for 3 reasons:

1) It locks the push/pull button. If after a previous fetch you could already push/pull, you now can't, and have to wait until the fetch is done. To reproduce this, go to a repo, await the fetch, and ensure there's something to push/pull. Then switch to another repo, and then switch back again and try to push/pull.

2) The spinning action is slightly distracting, as you eye naturally is drawn towards movement. While it's nice to know when the app is fetching, if it's doing it _all the time_ it becomes a bit of a nuisance and distraction.

3) A slight waste of resources (processing power, bandwidth, etc).

What do you think?

### PR DESCRIPTION
**Pull request by shiftkey**
Fixes #3329 for those who will switch repositories regularly by using the last fetched timestamp (which git updates using `.git/FETCH_HEAD` and we track in the app) and ignoring any background fetching for two minutes (chosen arbitrarily).

### CODE REVIEW

**Comment by shiftkey** [comment]
> I think this logic belongs in the `BackgroundFetcher` instead of here 🤔

That's where I initially started with this, but the dependency on `GitStore` for resolving `FETCH_HEAD` and how `AppStore` manages these instances lead me to moving it up.

I'll ponder on this some more, maybe there's some moving around we can do here...

**Comment by joshaber** [comment]
This is maybe a terrible idea, but it wouldn't bother me to store the last fetched time in `localStorage`. It means we wouldn't consider fetches from outside the app, but that doesn't seem like a serious downside.

Alternatively, `BackgroundFetcher` [already outsources](https://github.com/desktop/desktop/blob/930e9e8b62beaa2ffb1bae0bac09899c4b994666/app/src/lib/stores/helpers/background-fetcher.ts#L40) the actual fetch work. We could similarly outsource the "but should I really fetch?" work.

**Comment by shiftkey** [comment]
> We could similarly outsource the "but should I really fetch?" work.

I like this approach. Should be doable.

**Comment by shiftkey** [comment]
:apple:

File: app/src/lib/stores/app-store.ts [hunk 1]
```diff
@@ -889,6 +889,31 @@ export class AppStore {
...
+    const gitStore = this.getGitStore(repository)
+    const lastFetched = gitStore.lastFetched
+
+    if (!lastFetched) {
+      return true
+    }
+
+    const now = new Date()
+    const timeSinceFetch = now.getTime() - lastFetched.getTime()
+    const twoMinutes = 2 * 60 * 1000
```
Comment (joshaber) [inline app/src/lib/stores/app-store.ts]: What do think about pulling this into a top level constant, named something semantic like `BackgroundFetchMinimumInterval` or something?

**Comment by shiftkey** [comment]
🐕

**Pull request merged**